### PR TITLE
Added TypeExtensions

### PR DESCRIPTION
**Pull request by wintondeshong**
- GetPublicConstants<T>
- GetTypeName
- WhereWithAttribute<TAttribute>
- WhereWithoutAttribute<TAttribute>

### CODE REVIEW

File: src/AndcultureCode.CSharp.Extensions/AndcultureCode.CSharp.Extensions.md [hunk 1]
```diff
@@ -68,6 +68,12 @@
   - [ToEnumerable\`\`1(input,separator)](#M-AndcultureCode-CSharp-Extensions-StringExtensions-ToEnumerable``1-System-String,System-Char- 'AndcultureCode.CSharp.Extensions.StringExtensions.ToEnumerable``1(System.String,System.Char)')
   - [ToInt(number,defaultValue)](#M-AndcultureCode-CSharp-Extensions-StringExtensions-ToInt-System-String,System-Int32- 'AndcultureCode.CSharp.Extensions.StringExtensions.ToInt(System.String,System.Int32)')
   - [TryChangeType(value,conversionType)](#M-AndcultureCode-CSharp-Extensions-StringExtensions-TryChangeType-System-Object,System-Type- 'AndcultureCode.CSharp.Extensions.StringExtensions.TryChangeType(System.Object,System.Type)')
+- [TypeExtensions](#T-AndcultureCode-CSharp-Extensions-TypeExtensions 'AndcultureCode.CSharp.Extensions.TypeExtensions')
+  - [GetPublicConstantValues\`\`1()](#M-AndcultureCode-CSharp-Extensions-TypeExtensions-GetPublicConstantValues``1-System-Type- 'AndcultureCode.CSharp.Extensions.TypeExtensions.GetPublicConstantValues``1(System.Type)')
+  - [GetTypeName(obj)](#M-AndcultureCode-CSharp-Extensions-TypeExtensions-GetTypeName-System-Object- 'AndcultureCode.CSharp.Extensions.TypeExtensions.GetTypeName(System.Object)')
```
Comment (nhinhdao) [inline src/AndcultureCode.CSharp.Extensions/AndcultureCode.CSharp.Extensions.md:73]: Just wonder why `` needs to be there
image

File: test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        private class GetTypeNameTestStub { }
+
+        [Fact]
+        public void GetTypeName_Object_Overload_Given_Null_Returns_Null()
+        {
+            TypeExtensions.GetTypeName(null).ShouldBeNull();
+        }
+
+        [Fact]
+        public void GetTypeName_Object_Overload_Given_InstaceOf_Object_Returns_FullName()
```
Comment (nhinhdao) [inline test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs:67]: Typo `instance`

File: test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+            // Act
+            var result = type.GetTypeName();
+
+            // Assert
+            result.ShouldContain(typeof(GetTypeNameTestStub).FullName);
+        }
+
+        [Fact]
+        public void GetTypeName_Object_Overload_Given_InstaceOf_Object_Returns_AssemblyQualifiedName()
```
Comment (nhinhdao) [inline test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs:80]: Typo here too

File: test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            var result = sut.WhereWithoutAttribute<AttributeUsageAttribute>();
+
+            // Assert
+            result.ShouldNotBeNull();
+            result.Count.ShouldBe(1);
+            result.ShouldContain(expected);
+        }
+
+        [Fact]
+        public void WhereWithoutAttribute_Given_Mixed_Types_With_Attribute_Returns_Undecorated_Types_Without_Duplicates()
```
Comment (nhinhdao) [inline test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs:239]: Line break glitch

Comment (nhinhdao) [reply test/AndcultureCode.CSharp.Extensions.Tests/TypeExtensionsTests.cs:239]: Oh nvm, it is not an issue from this view

File: src/AndcultureCode.CSharp.Extensions/AndcultureCode.CSharp.Extensions.md (hunk 1 above)
Comment (wintondeshong) [reply src/AndcultureCode.CSharp.Extensions/AndcultureCode.CSharp.Extensions.md:73]: Yeah, odd way generics are represented

**Pull request merged**